### PR TITLE
Enable autograd graph to propagate after multi-device syncing for loss functions in `ddp`

### LINKED ISSUE #2745: Autograd with DDP
I have a setup with `torch.Lightning` where I'm using custom `torchmetrics.Metric` as loss function contributions. Now I want to be able to do it with `ddp` by setting `dist_sync_on_step=True`, but the gradients are not propagated during the `all_gather`. All I want is for the tensor on the current process to have its autograd graph kept for the backward pass after the syncing operations. I've only just began looking into distributed stuff in `torch`, so I'm not experienced in these matters. But following the `forward()` call of `Metric` (at each training batch step), it then calls `_forward_reduce_state_update()`, which calls the `compute()` function wrapped by `_wrap_compute()`, which would do `sync()`, which finally calls `_sync_dist()`. And it looks like the syncing uses `torchmetrics.utilities.distributed.gather_all_tensors`.

 I just wanted to ask if it is possible to achieve what I want by modiyfing `_simple_gather_all_tensors` ([here](https://github.com/Lightning-AI/torchmetrics/blob/77065a503dd22d273970a72ebe62b507a1d4db28/src/torchmetrics/utilities/distributed.py#L91))? `_simple_gather_all_tensors` presented here for reference.
```py
def _simple_gather_all_tensors(result: Tensor, group: Any, world_size: int) -> List[Tensor]:
    gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
    torch.distributed.all_gather(gathered_result, result, group)
    return gathered_result
```
I'm guessing that `result` still carries the autograd graph. My naive hope is that we can just update `gathered_result` with the input `result` (carrying the autograd graph) to achieve the desired effect.

For context, my use case is such that batches can have very inhomogeneous `numel`s, so each device could have error tensors with very different `numel`s such that taking a mean of `MeanSquaredError`s may not be ideal. Ideally, if the syncing holds the autograd graph, the per-step loss would be the "true" metric as per its definition and the gradients would be consistent with that definition (so syncing is done once for for each loss metric contribution, and once for the backward at each training step, I think).

Thank you!

### PR DESCRIPTION
**Pull request by cw-tan**
## What does this PR do?

Fixes #2745

Single-line enhancement proposed in #2745, that is, to enable the propagation of the autograd graph after the `all_gather` operation. This is useful for syncing loss functions in a `ddp` setting. 

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2754.org.readthedocs.build/en/2754/

### CODE REVIEW

**Comment by Borda** [comment]
That sounds good to me, but can we add a test for this enhancement?

**Comment by cw-tan** [comment]
> That sounds good to me, but can we add a test for this enhancement?

Thanks for the prompt response @Borda.

I'm thinking that `_test_ddp_gather_uneven_tensors` ([here](https://github.com/Lightning-AI/torchmetrics/blob/748caee4e21d7ce9a97ed2612363e12eac94356f/tests/unittests/bases/test_ddp.py#L63)) and `_test_ddp_gather_uneven_tensors_multidim` ([here](https://github.com/Lightning-AI/torchmetrics/blob/748caee4e21d7ce9a97ed2612363e12eac94356f/tests/unittests/bases/test_ddp.py#L71C5-L71C45)) in `tests/unittests/bases/test_ddp.py` already cover the correctness of `gather_all_tensors`. I'm not sure what other `ddp` tests there are, but those tests should help tell us if the change I made isn't breaking existing functionality. Let me know if you had something else in mind for this.

I can make an additional unittest in `tests/unittests/bases/test_ddp.py` to give a tensor that `requires_grad` to `gather_all_tensors`, compute some scalar from them (proxy for a loss), and compute grads two ways (one going through the `all_gather`, one that doesn't) and compare. So this tests that the change achieves the desired effect. How does that sound?

**Comment by Borda** [comment]
> I can make an additional unittest in `tests/unittests/bases/test_ddp.py` to give a tensor that `requires_grad` to `gather_all_tensors`, compute some scalar from them (proxy for a loss), and compute grads two ways (one going through the `all_gather`, one that doesn't) and compare. So this tests that the change achieves the desired effect. How does that sound?

yeah, that sounds good to me :)

**Comment by cw-tan** [comment]
Update: to accommodate both cases where tensors from different ranks have the same/different shape, the line to put the original tensor (holding the AD graph) back into the gathered list was added in two places in the code. 

Because of the two cases, I wrote two unittests to account for each. Interestingly, both pass `2.X stable`, but for `1.X LTS`, the "same shape" test passes but "different shape" test fails, and for `1.10 oldest`, the "different shape" test passes but "same shape" test fails😅.   I'll double check for bugs, but the actual code change is just two lines (and all other tests pass, so existing functionality still works), and the unittests are pretty short. The dependency of the unittests passing on different `torch` versions seems to indicate that it might be  a `torch` versioning issue, maybe to do with `ddp` behavior? Any thoughts, @Borda ?

**Comment by Borda** [comment]
> I wrote two unittests to account for each. Interestingly, both pass `2.X stable`, but for `1.X LTS`, the "same shape" test passes but "different shape" test fails, and for `1.10 oldest`, the "different shape" test passes but "same shape" test fails😅.

that is strange and worse some more investigation...
cc: @SkafteNicki

File: src/torchmetrics/utilities/distributed.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def class_reduce(
 def _simple_gather_all_tensors(result: Tensor, group: Any, world_size: int) -> List[Tensor]:
     gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
     torch.distributed.all_gather(gathered_result, result, group)
+    gathered_result[torch.distributed.get_rank(group)] = result
```
[the remark this answers is not in the document]
Comment (cw-tan) [reply src/torchmetrics/utilities/distributed.py:95]: added comment  `# to propagate autograd graph from local rank (achieves intended effect for torch> 2.0)`

File: src/torchmetrics/utilities/distributed.py [hunk 2]
```diff
@@ -144,4 +145,5 @@ def gather_all_tensors(result: Tensor, group: Optional[Any] = None) -> List[Tens
     for idx, item_size in enumerate(local_sizes):
         slice_param = [slice(dim_size) for dim_size in item_size]
         gathered_result[idx] = gathered_result[idx][slice_param]
+    gathered_result[torch.distributed.get_rank(group)] = result
```
[the remark this answers is not in the document]
Comment (cw-tan) [reply src/torchmetrics/utilities/distributed.py:150]: added comment  `# to propagate autograd graph from local rank (achieves intended effect for torch> 2.0)`

File: tests/unittests/bases/test_ddp.py [hunk 3]
```diff
@@ -77,6 +77,40 @@ def _test_ddp_gather_uneven_tensors_multidim(rank: int, worldsize: int = NUM_PRO
         assert (val == torch.ones_like(val)).all()
 
 
+def _test_ddp_gather_autograd_same_shape(rank: int, worldsize: int = NUM_PROCESSES) -> None:
+    tensor = torch.ones(50, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (cw-tan) [reply tests/unittests/bases/test_ddp.py:89]: Added docstring and modification `@pytest.mark.skipif(not _TORCH_GREATER_EQUAL_2_1, reason="test only works on newer torch versions")` as per your comment on PT versions. (I just used the 2.1 flag because it was already imported in the same file -- could change it if you want)

File: tests/unittests/bases/test_ddp.py [hunk 4]
```diff
@@ -77,6 +77,40 @@ def _test_ddp_gather_uneven_tensors_multidim(rank: int, worldsize: int = NUM_PRO
...
+        else:
+            scalar1 = scalar1 + torch.sum(result[idx] * torch.ones_like(result[idx]))
+        scalar2 = scalar2 + torch.sum(result[idx] * torch.ones_like(result[idx]))
+    gradient1 = torch.autograd.grad(scalar1, [tensor], retain_graph=True)[0]
+    gradient2 = torch.autograd.grad(scalar2, [tensor])[0]
+    assert torch.allclose(gradient1, gradient2)
+
+
+def _test_ddp_gather_autograd_different_shape(rank: int, worldsize: int = NUM_PROCESSES) -> None:
+    tensor = torch.ones(rank + 1, 2 - rank, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (cw-tan) [reply tests/unittests/bases/test_ddp.py:114]: same as above

File: src/torchmetrics/utilities/distributed.py [hunk 5]
```diff
@@ -144,4 +146,6 @@ def gather_all_tensors(result: Tensor, group: Optional[Any] = None) -> List[Tens
     for idx, item_size in enumerate(local_sizes):
         slice_param = [slice(dim_size) for dim_size in item_size]
         gathered_result[idx] = gathered_result[idx][slice_param]
+    # to propagate autograd graph from local rank (achieves intended effect for torch> 2.0)
+    gathered_result[torch.distributed.get_rank(group)] = result
```
Comment (SkafteNicki) [inline src/torchmetrics/utilities/distributed.py:150]:
```suggestion
    if _TORCH_GREATER_EQUAL_2_1:
        gathered_result[torch.distributed.get_rank(group)] = result
```

**Review by SkafteNicki — COMMENTED** [review]
seeems the two test functions are now included twice in the `test_ddp.py` file, please check

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
> > seeems the two test functions are now included twice in the test_ddp.py file, please check
> 
> ah yes, sorry about that -- probably left behind when I tried rebasing and force-pushing with the pre-commit CI commits.
> 
> Thanks for the review and changes @SkafteNicki. Unfortunately, it looks like the 2.X stable tests are failing now. This may suggest that something more subtle was happening with the failure of the torch < 2.0 tests earlier?

That is strange, but yes @cw-tan it may very well be the case that this is also what caused the error in the older Pytorch versions.
I will try to find some time to debug but feel free to also take a stab at this new issue

**Comment by SkafteNicki** [comment]
@cw-tan this is really strange, I am trying to debug this locally and I am seeing that the tests are failing at random. Eg. if I run them 10 times in a row I get a output from pytest like this:

```
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_same_shape-1-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_same_shape-3-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_same_shape-7-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_same_shape-9-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_different_shape-5-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_different_shape-6-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_different_shape-8-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_different_shape-9-10] - AssertionError
FAILED tests/unittests/bases/test_ddp.py::test_ddp_autograd[_test_ddp_gather_autograd_different_shape-10-10] - AssertionError
```
with 4/10 of the "same shape" tests failing and 5/10 of the "different shape" test failing. But I cannot see there is any randomization going on in the tests?

**Comment by cw-tan** [comment]
@SkafteNicki indeed this is an odd one. Though adding the `with torch.no_grad():` in my recent commit only had the "different shape" test failing -- were both "same shape" and "different shape" tests failing before? I'm thinking maybe covering more of the code with `torch.no_grad()` except for the parts we want for the autograd graph to be propagated might be worth a try. Though I don't actually know why it would help a priori.

**Comment by cw-tan** [comment]
@SkafteNicki sorry for the mess, I'm just trying to use the CI tests on all torch versions again but hopefully incorporating several trials (to check for indeterminism) and with the `no_grad` changes I made.

**Comment by SkafteNicki** [comment]
> @SkafteNicki sorry for the mess, I'm just trying to use the CI tests on all torch versions again but hopefully incorporating several trials (to check for indeterminism) and with the `no_grad` changes I made.

@cw-tan That is completely okay whatever it takes to debug the issue. If you want to locally to run tests multiple times i recommend installing:
https://pypi.org/project/pytest-repeat/
and then running `pytest` command with
```pytest --count=X```
for X repeating evaluations

**Comment by SkafteNicki** [comment]
Alright, I finally sat down to understand what was going on here. The non-deterministic behavior was really strange to me, so I tried a lot of debugging and realized that order of output of the `all_gather` operation was not consistent. This meant that sometimes process 0 was the first element in the gathered list, sometimes the second e.g.
```
Process 0 gathered tensors: [tensor([0.]), tensor([1.])]  # rank and gathering order matches
Process 1 gathered tensors: [tensor([0.]), tensor([1.])]
Process 0 gathered tensors: [tensor([1.]), tensor([0.])]  # rank and gathering order do not match
Process 1 gathered tensors: [tensor([1.]), tensor([0.])]
```
I found out that the reason for this is that when the `setup` function (that initializes the ddp group) is called with another `pool.map/pool.starmap` than where the test is, this can lead to the processes not matches the expected order. This is exactly what we do in the test code currently
https://github.com/Lightning-AI/torchmetrics/blob/abdd2c4e6cbf72741b162dd6361a6c1024df69a2/tests/unittests/conftest.py#L66-L68
and then `pytest.pool.starmap` is called when we want to run a test in ddp mode. The solution was to call the `setup` function during the test function and everything is in the expected order. See this commit for details: https://github.com/Lightning-AI/torchmetrics/pull/2754/commits/48e699b6dfaa7a5fe46d8a984722f155c510c4d5.
This has not been a problem before because we normally reduce all the states in some way e.g. sum them and then the order does not matter at all. Hopefully, this also means that it works regardless of Pytorch version.

@cw-tan sorry for the headache this must have been to debug. I have significantly simplified the tests you had for me to understand what was going on. Hope this still is fine with you.

**Comment by cw-tan** [comment]
@SkafteNicki Fantastic, thank you so much! I'm just excited to see this feature released so I can remove the monkeypatch in my own code to achieve the same effects. I think the docs are the remaining change -- still some details about only being PyTorch 2 compatible, but the tests have passed for other versions.

File: docs/source/pages/overview.rst [hunk 6]
```diff
@@ -492,6 +492,10 @@ In practice this means that:
 
 A functional metric is differentiable if its corresponding modular metric is differentiable.
 
+For PyTorch versions 2.1 or higher, differentiation in DDP mode is enabled, allowing autograd graph
+propagation after the ``all_gather`` operation. This is useful for synchronizing metrics used as
+loss functions in a DDP setting.
```
Comment (Borda) [inline docs/source/pages/overview.rst:497]:
```suggestion
.. caution::
    For PyTorch versions 2.1 or higher, differentiation in DDP mode is enabled, allowing autograd graph
    propagation after the ``all_gather`` operation. This is useful for synchronizing metrics used as
    loss functions in a DDP setting.
```

**Review by Borda — APPROVED** [review]
